### PR TITLE
Streamline `rustc_driver_impl` pretty-printing.

### PR DESCRIPTION
**Pull request by nnethercote**
This PR simplifies a lot of unnecessary structure in
`rustc_driver_impl/src/pretty.rs`. It removes some traits and functions,
simplifies some structs, renames some things for increased consistency, and
eliminates some boilerplate code. Overall it cuts more than 150 lines of code.

r? @compiler-errors

### CODE REVIEW

**Comment by nnethercote** [comment]
This PR is probably best reviewed one commit at a time. Each commit is a self-contained change, but note that the progress through the commits is not entirely a straight line; I was feeling my way as I went. Some things in earlier commits are effectively undone in later commits (and the log messages indicate some of these). If I did it again from scratch I would end up with slightly fewer commits, but I don't think that's worth the effort.

**Comment by nnethercote** [comment]
I fixed the conflicts.

File: compiler/rustc_driver_impl/src/pretty.rs [hunk 1]
```diff
@@ -18,24 +18,27 @@ pub use self::PpMode::*;
...
 
-impl<'tcx> pprust_ast::PpAnn for NoAnn<'tcx> {}
-impl<'tcx> pprust_hir::PpAnn for NoAnn<'tcx> {
+impl<'tcx> pprust_hir::PpAnn for HirNoAnn<'tcx> {
     fn nested(&self, state: &mut pprust_hir::State<'_>, nested: pprust_hir::Nested) {
-        if let Some(tcx) = self.tcx {
-            pprust_hir::PpAnn::nested(&(&tcx.hir() as &dyn hir::intravisit::Map<'_>), state, nested)
-        }
+        pprust_hir::PpAnn::nested(
+            &(&self.tcx.hir() as &dyn hir::intravisit::Map<'_>),
```
Comment (compiler-errors) [inline compiler/rustc_driver_impl/src/pretty.rs:32]: This is an unnecessary `&`?

File: compiler/rustc_driver_impl/src/pretty.rs [hunk 2]
```diff
@@ -360,119 +213,119 @@ fn write_or_print(out: &str, sess: &Session) {
...
-            format!("{krate:#?}")
-        }
-        _ => unreachable!(),
-    };
+    }
 
-    write_or_print(&out, sess);
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        match self {
+            PrintExtra::AfterParsing { .. } => panic!("PrintExtra::tcx"),
```
Comment (compiler-errors) [inline compiler/rustc_driver_impl/src/pretty.rs:236]:
```suggestion
            PrintExtra::AfterParsing { .. } => bug!("PrintExtra::tcx"),
```

**Review by compiler-errors — APPROVED** [review]
r=me after nits or not

File: compiler/rustc_driver_impl/src/pretty.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_driver_impl/src/pretty.rs:32]: It's necessary, if I remove it I get this:
```
error[E0277]: the trait bound `dyn rustc_hir::intravisit::Map<'_>: rustc_hir_pretty::PpAnn` is not satisfied
  --> compiler/rustc_driver_impl/src/pretty.rs:32:35
   |
32 |         pprust_hir::PpAnn::nested(&self.tcx.hir() as &dyn hir::intravisit::Map<'_>, state, nested)
   |         ------------------------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `rustc_hir_pretty::PpAnn` is not implemented for `dyn rustc_hir::intravisit::Map<'_>`
   |         |
   |         required by a bound introduced by this call
   |
   = help: the trait `rustc_hir_pretty::PpAnn` is implemented for `&dyn rustc_hir::intravisit::Map<'_>`
```

File: compiler/rustc_driver_impl/src/pretty.rs (hunk 2 above)
Comment (nnethercote) [reply compiler/rustc_driver_impl/src/pretty.rs:236]: :+1:

**Pull request merged**